### PR TITLE
Use TestClock when TestStream is present in pipeline

### PR DESCRIPTION
**Pull request by charlesccychen**
This fixes an issue where TestClock was incorrectly not used when a TestStream was used in a pipeline.

### CODE REVIEW

**Comment by charlesccychen** [comment]
R: @robertwb 
CC: @mariapython

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 1]
```diff
@@ -327,25 +327,45 @@ def __init__(self):
...
+    from apache_beam.testing.test_stream import TestStream
+
+    # Performing configured PTransform overrides.
+    pipeline.replace_all(_get_transform_overrides(pipeline.options))
+
+    # If the TestStream I/O is used, use a mock test clock.
+    class _TestStreamUsageVisitor(PipelineVisitor):
+      """For internal use only; no backwards-compatibility guarantees.
+
+      Visitor for determining whether a Pipeline can be run on the FnApiRunner.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: Update the docstring.

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 2]
```diff
@@ -327,25 +327,45 @@ def __init__(self):
...
     from apache_beam.runners.direct.transform_evaluator import \
       TransformEvaluatorRegistry
+    from apache_beam.testing.test_stream import TestStream
+
+    # Performing configured PTransform overrides.
+    pipeline.replace_all(_get_transform_overrides(pipeline.options))
+
+    # If the TestStream I/O is used, use a mock test clock.
+    class _TestStreamUsageVisitor(PipelineVisitor):
+      """For internal use only; no backwards-compatibility guarantees.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: This isn't needed for private classes.

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 3]
```diff
@@ -327,25 +327,45 @@ def __init__(self):
...
+
+      Visitor for determining whether a Pipeline can be run on the FnApiRunner.
+      """
+
+      def __init__(self):
+        self.uses_test_stream = False
+
+      def visit_transform(self, applied_ptransform):
+        transform = applied_ptransform.transform
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: Nit: I'd just write

`if isinstance(applied_ptransform.transform, TestStream)`

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 1 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Done.

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 2 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Done.

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 3 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Done.

**Comment by charlesccychen** [comment]
Thanks, PTAL.

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 4]
```diff
@@ -327,25 +324,41 @@ def __init__(self):
...
+
+      def __init__(self):
+        self.uses_test_stream = False
+
+      def visit_transform(self, applied_ptransform):
+        if isinstance(applied_ptransform.transform, TestStream):
+          self.uses_test_stream = True
+
+    visitor = _TestStreamUsageVisitor()
+    pipeline.visit(visitor)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: This pattern feels "impure" to me: create a visitor, visit, check a bit on the visitor. I suppose we could put a method visit_pipeline() on _TestStreamUsageVisitor that calls pipeline.visit(self) and returns self.uses_test_stream.

Comment (robertwb) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: I was just commenting that querying whether a a test stream is present requires three lines and reaching into a private variable. Instead, add

```
def visit_pipeline(self, pipeline):
  self._uses_test_stream = False
  pipeline.visit(self)
  return self._uses_test_stream
```

and then you can write 

`use_test_clock = _TestStreamUsageVisitor().visit_pipeline(pipeline)`

Comment (mariapython) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: > **robertwb** wrote:
> I was just commenting that querying whether a a test stream is present requires three lines and reaching into a private variable. Instead, add
> 
> ```
> def visit_pipeline(self, pipeline):
>   self._uses_test_stream = False
>   pipeline.visit(self)
>   return self._uses_test_stream
> ```
> 
> and then you can write 
> 
> `use_test_clock = _TestStreamUsageVisitor().visit_pipeline(pipeline)`


Sorry, it was not in response to your message, but to Charles as a general comment, that as is, the self._use_test_clock = False is still in the direct_runner.py

Comment (charlesccychen) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: > **mariapython** wrote:
> Sorry, it was not in response to your message, but to Charles as a general comment, that as is, the self._use_test_clock = False is still in the direct_runner.py


Thanks, I removed the redundant property.

Comment (charlesccychen) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Fixed, thanks.

**Comment by charlesccychen** [comment]
Rebased to head and tests pass. Can you PTAL? Thanks!

**Pull request merged**